### ISSUE TITLE
add healthcheck instruction

### ISSUE THREAD

**Issue by davidkarlsen** [elastic/elasticsearch-docker#60]
## Feature Description

It would be nice to add the HEALTCHECK feature, see: https://docs.docker.com/engine/reference/builder/ to deduct ES is up and running in a "healthy" state. Healthy in this context should probably mean available, and not cluster state yellow etc

**Comment by dliappis** [comment]
Just wanted to provide some feedback on this one.

I had a look and came up with a reasonable implementation. However I realized that there are more complications than initially thought:

1. If the user redefines [http.port](https://www.elastic.co/guide/en/elasticsearch/reference/current/modules-http.html) the healthcheck script needs to apply heuristics to check for the port either in env vars, and/or definitions in `elasticsearch.yml`.\[1\] Such heuristics won't help if the user overrides CMD with something like `bin/elasticsearch -Ehttp.port=9900` (configuration option D. in our docs).

2. It's also possible that [xpack.security.ssl.enabled=true](https://www.elastic.co/guide/en/x-pack/current/ssl-tls.html) has been set, in which case the healthcheck script needs to either perform similar heuristics as in 1, or check for both http and then https if the http fails. The latter is not attractive because it pollutes the logs with failed attempts depending on the value of `--interval`.

3. Kubernetes has liveness and readiness probes that currently are overlapping HEALTHCHECK. As a result in https://github.com/kubernetes/kubernetes/issues/50703 Kubernetes anyway disables any `HEALTHCHECK`.

4. Healthcheck commands can also be defined using [docker-compose](https://docs.docker.com/compose/compose-file/#healthcheck). This, sounds more sane IMO, as the user can configure a simple curl command like `curl -f http://localhost:theport || exit 1` and, optionally, tweak http/https and specify username/password and provide similar functionality to the liveness/readiness k8s probes.

So to summarize, I think with the amount of configurability that Elasticsearch and the docker image presents, it's not prudent to add brittle and complicated logic to a healthcheck script, when the same functionality can be achieved with a docker-compose `healthcheck:` or liveness/readiness k8s probes, without the guesswork.

\[1\] We could also inspect the output of `netstat -nalp`, but there's no warranty that the `HEALTHCHECK` script runs as root e.g. on Openshift.

cc @jarpy

**Comment by davidkarlsen** [comment]
Good points. Thank you for the nice summary!
